### DISCUSSION TITLE
32-bit ATN Serialization

### DISCUSSION THREAD

**Discussion by parrt** [antlr/antlr4#3580] [Ideas]
Hi @ericvergnaud @jcking @marcospassos @mike-lischke @KvanTTT!

Now that all targets other than java use integer arrays, perhaps it's time to move away from the 16 bit encoding used now in favor of 32 bit integers. ThatGives the advantage that ATNs are not limited in size, and there are a number of bugs issued for this. Java would have to be treated specially but the idea is simple: remove the masking of numbers with 0xFFFF and avoid the use of 2 16-bit numbers represent 32-bit values.

@KvanTTT  has started on various variations of this but a new PR with his 32-bit ATN unit tests shouldn't be too hard.  this would require a small change to each target deserialization to use int32 not int16 (although I think most to just you I'm integer but Go using uint16) and avoid packing 2 16s into a 32 bit num for unicode>16bit.

 shall we do this for 4.10? while we are doing work on the serialization seems like we should fix this and then stop thinking about it.

**Comment by ericvergnaud** [comment]
Sure.

May I suggest starting by creating a test that does involve 32-bits ?

Envoyé de mon iPhone

**Reply by parrt** [reply]
I was thinking the same. @KvanTTT  Created some of these already in one of his PRs.

**Reply by KvanTTT** [reply]
Yes, it exists here: https://github.com/antlr/antlr4/pull/3546/files#diff-859a1117c3f13396162959e824858c01d179a61f3280ab64fd68c9c9490b2729R71

**Comment by mike-lischke** [comment]
@parrt Can you give a bit more context? I don't know what you mean. Maybe point to a place in the code to allow following the discussion?

**Reply by parrt** [reply]
Sure. ANTLR creates a state machine (ATN) from the user's grammar that is used to make parsing decisions at parse time, as you know. In order to get that state machine into the generated code, we need to serialize the state machine into a series of integers that represent the various states and transitions with their labels. Here is how the tool does the serialization:

https://github.com/antlr/antlr4/blob/master/runtime/Java/src/org/antlr/v4/runtime/atn/ATNSerializer.java#L67

The end result of the serialization is a list of 32-bit integers with a few bits of encoding. E.g., [EOF](https://github.com/antlr/antlr4/blob/master/runtime/Java/src/org/antlr/v4/runtime/atn/ATNSerializer.java#L150) and any -1 values (such as ["no rule"](https://github.com/antlr/antlr4/blob/master/runtime/Java/src/org/antlr/v4/runtime/atn/ATNSerializer.java#L313) or "no lexer action") are encoded as the maximum 16-bit char, unsigned 0xFFFF.  Also 32-bit unicode edge labels are encoded as [two 16-bit values](https://github.com/antlr/antlr4/blob/master/runtime/Java/src/org/antlr/v4/runtime/atn/ATNSerializer.java#L169) (the only such 32-bit -> 16-bit encoding). This is all despite the fact that everything is still kept as a list of 32 bit integers.

From this list of integers that are all technically uint16, we generate an integer array for all the targets except java. Java does not properly handle static initialized integer arrays and so we have to fake it with a string of 16bit unicode. This prevents truly large ATNs from being serialized for any target because the serialization process fundamentally limits the size of the integers. 

**Proposal**. I propose that we convert the `serialize()` method to simply create a list of 32 signed integers so that -1 is -1 not 0xFFFF and 32-bit unicode edge labels are not 16-bit encoded.   This will simplify the code during deserialization. E.g.,Here is a [decoding of 0xFFFF as -1](https://github.com/antlr/antlr4/blob/master/runtime/Java/src/org/antlr/v4/runtime/atn/ATNDeserializer.java#L273). And then this special case [unicode sets code](https://github.com/antlr/antlr4/blob/master/runtime/Java/src/org/antlr/v4/runtime/atn/ATNDeserializer.java#L313) goes away.  We will need to keep the Java target as a special case but that's no big deal.

I believe @KvanTTT  already has unit tests too check for the special cases and also has unit tests to validate ATN serialized as thirty two bit integers; I just can't find them at the moment easily.

**Strategy**.  Create a general `serialize()` method that does no encoding and returns a list of 32 bits signed integers.  Sharing as much as possible, we create a specialized version to serialize things for java. The Java target deserializer would stay the same but other targets would remove the parts that handle 32bit char sets specially; e.g., [for Go target](https://github.com/antlr/antlr4/blob/master/runtime/Go/antlr/atn_deserializer.go#L101). Any uint16 types would need to be converted to just int, but that is likely only Go and C++ targets. We also need to pull in @KvanTTT's tests. The conversion process is likely simple enough that even somebody like me, who was not an expert in all of these languages, could do the transformation. (I just managed to get rid of the string -> int deserialization portions for a few of the targets.)

**Reply by parrt** [reply]
@KvanTTT Life is a bit complicated for you at the moment, but is this something you are interested in handling?

**Reply by mike-lischke** [reply]
Aha, right, I remember that part of the code. Can we change it also so that we get more type safety? Currently it's just a bunch of numbers, that are arbitrarily interpreted to be a count or state number etc. Instead that should be a stream of structures with well known fields and their types.

A possible implementation could be like the PNG chunk structure: there's base record in each chunk with the overall length of that chunk and an indicator what type of chunk that is. The good thing about this is you can place the chunks in any order and you can easily add new chunks at any time without using a UUID or version field and so on. Implementations that don't know how to handle a chunk can just jump over it (because of that length field).

**Reply by parrt** [reply]
A very interesting idea @mike-lischke!   I suppose that we could create java objects that represent this serialized form (though course we all already have the full ATN as a set of state objects), and then the individual target templates could generate C++ `struct`s or whatever.  Many targets may not have such useful critters as `struct`s though.

Given that I hope we don't have to mess with the ATN serialization ever again, the fastest and lowest risk solution is to tweak what we have. I will make an attempt and then have you folks take a look but it should not be very hard.

**Reply by KvanTTT** [reply]
> A possible implementation could be like the PNG chunk structure

I afraid it could increase the size of output serialized data because of types serialization. Moreover there is no compression for ATN unlike PNG.

**Reply by mike-lischke** [reply]
> Given that I hope we don't have to mess with the ATN serialization ever again, the fastest and lowest risk solution is to tweak what we have. I will make an attempt and then have you folks take a look but it should not be very hard.

OK, if that's a one time thing then this is probably the way to go.

**Reply by lppedd** [reply]
>  Java does not properly handle static initialized integer arrays

Where can I find more info on this topic? I didn't know there was an issue about static initialization.

@ericvergnaud found this discussion by following what you told me yesterday, thanks!

**Comment by KvanTTT** [comment]
32-bit integers increases the size twice of output ATN compare to 16-bit. Is this ok?

> shall we do this for 4.10? while we are doing work on the serialization seems like we should fix this and then stop thinking about it.

I think so becuase in 4.10 we've changed ATN format anyway.

**Reply by parrt** [reply]
Yep,It will increase the size but based upon the numbers we've seen it'll be like 12k -> 24k for a big grammar like SQL. that's probably a reasonable trade to get a simplified atn that doesn't do any encoding at all.   if this becomes an issue in the future coma we can always add the high-bit encoding trick.  I like the idea of using -1 instead of trying to encode Token.EOF as 0xFFFF or something as well.

**Reply by parrt** [reply]
I'm going to try building this today

**Reply by parrt** [reply]
Almost done!!!

**Reply by parrt** [reply]
See https://github.com/antlr/antlr4/pull/3591

**Comment by parrt** [comment]
Wow. Ok, Java now supports 32-bit ATNs. It handles @KvanTTT's large generated lexer and the case where the token type is 0xFFFF. :)   the remaining issue I think, other than cleanup you guys suggest, is to handle the shift by two thing for java so that the modified utf-8 string encoding in the class files does not make 0 and 1 inefficient to represent.  I think the easiest answer is to do the bump by 2 trick unless the 32-bit encoding kicks in. See https://github.com/antlr/antlr4/pull/3591

**Reply by mike-lischke** [reply]
Can I see that large generated Lexer @KvanTTT? I'd like to compare it with stuff I have.

**Reply by parrt** [reply]
i emailed it to you

**Reply by KvanTTT** [reply]
Sure, sorry for delay, it's also available on gist: https://gist.github.com/KvanTTT/301b9d4d97e8b0a74ab1a7d6c0dee715 (or can be generated using `getAtnStatesSizeMoreThan65535Descriptor` test)

**Comment by parrt** [comment]
I'm thinking That the bump by two feature isn't really a big problem. Yes it creates a bigger class file but that has all kinds of symbol table stuff and other junk as well. when it is loaded by some fast C code, it unpacks the utf8 into 16bit char strings so we don't have a memory problem there.  I think I will carve out the bump by 2 stuff. we can revise it later if for some reason it's a problem
